Index: src/main/java/org/exoplatform/services/jcr/ext/audit/AuditServiceImpl.java
===================================================================
--- src/main/java/org/exoplatform/services/jcr/ext/audit/AuditServiceImpl.java	(revision 74200)
+++ src/main/java/org/exoplatform/services/jcr/ext/audit/AuditServiceImpl.java	(working copy)
@@ -677,7 +677,7 @@
       ItemData storage = session.getTransientNodesManager().getItemData(AUDIT_STORAGE_ID);
 
       if (storage == null) {
-        SessionChangesLog changesLog = new SessionChangesLog(session.getId());
+        SessionChangesLog changesLog = new SessionChangesLog(session);
 
         // here should be added to TransactionalDataManager (i.e. saved
         // immediatelly!
